Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift

```diff
@@ -303,7 +313,17 @@ public final class JSONRPCConnection {
     send(_rawData: dispatchData)
   }
 
-  func send(encoding: (JSONEncoder) throws -> Data) {
+  private func sendMessageSynchronously(_ messageData: Data) {
+    let header = "Content-Length: \(messageData.count)\r\n\r\n"
+    guard let headerData = header.data(using: .utf8) else {
+      log("error encoding header", level: .error)
+      return
+    }
+    let stdOutHandle = FileHandle(fileDescriptor: sendIO.fileDescriptor)
```

Comment:
This seems unsafe since there may be other asynchronous writers.  How about something like this?

```swift
let semaphore = DispatchSemaphore(value: 0)
sendIO.barrier {
  ... write to the file descriptor, which we now have exclusive access to ...
  semaphore.signal()
}

_ = semaphore.wait(timeout: .now() + .seconds(3))
```

Although perhaps the specific timeout should be passed in.

Reply:
Tried to address this in https://github.com/apple/sourcekit-lsp/pull/334/commits/3f4c1c5c29a509809ab18d405b2ea0a7176afea7 without the `barrier` and preferring existing means ... I hope the `completionHandler` is ok?